homeassistant/components/modbus/validators.py

```diff
@@ -354,7 +219,132 @@ def register_int_list_validator(value: Any) -> Any:
 
 def check_config(config: dict) -> dict:
     """Do final config check."""
-    config2 = duplicate_modbus_validator(config)
-    config3 = scan_interval_validator(config2)
-    config4 = duplicate_entity_validator(config3)
-    return config4
+    hosts: set[str] = set()
+    hub_names: set[str] = set()
+    hub_name_inx = 0
+    minimum_scan_interval = 0
+    ent_names: set[str] = set()
+    ent_addr: set[str] = set()
+
+    def validate_modbus(hub: dict, hub_name_inx: int) -> bool:
+        """Validate modbus entries."""
+        host: str = (
+            hub[CONF_PORT]
+            if hub[CONF_TYPE] == SERIAL
+            else f"{hub[CONF_HOST]}_{hub[CONF_PORT]}"
+        )
+        if CONF_NAME not in hub:
+            hub[CONF_NAME] = (
+                DEFAULT_HUB if not hub_name_inx else f"{DEFAULT_HUB}_{hub_name_inx}"
+            )
+            hub_name_inx += 1
+            err = f"Modbus host/port {host} is missing name, added {hub[CONF_NAME]}!"
+            _LOGGER.warning(err)
+        name = hub[CONF_NAME]
+        if host in hosts or name in hub_names:
+            err = f"Modbus {name} host/port {host} is duplicate, not loaded!"
+            _LOGGER.warning(err)
+            return False
+        hosts.add(host)
+        hub_names.add(name)
+        return True
+
+    def validate_entity(
+        hub_name: str,
+        entity: dict,
+        minimum_scan_interval: int,
+        ent_names: set,
+        ent_addr: set,
+    ) -> bool:
+        """Validate entity."""
+        name = entity[CONF_NAME]
+        addr = str(entity[CONF_ADDRESS])
+        scan_interval = entity.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+        if scan_interval < 5:
+            _LOGGER.warning(
+                (
+                    "%s %s scan_interval(%d) is lower than 5 seconds, "
+                    "which may cause Home Assistant stability issues"
+                ),
+                hub_name,
+                name,
+                scan_interval,
+            )
+        entity[CONF_SCAN_INTERVAL] = scan_interval
+        minimum_scan_interval = min(scan_interval, minimum_scan_interval)
+        for conf_type in (
+            CONF_INPUT_TYPE,
+            CONF_WRITE_TYPE,
+            CONF_COMMAND_ON,
+            CONF_COMMAND_OFF,
+        ):
+            if conf_type in entity:
+                addr += "_" + str(entity[conf_type])
+        inx = entity.get(CONF_SLAVE, None) or entity.get(CONF_DEVICE_ADDRESS, 0)
+        addr += "_" + str(inx)
+        loc_addr: set[str] = set()
+        loc_addr.add(addr)
+
+        if CONF_TARGET_TEMP in entity:
+            a = str(entity[CONF_TARGET_TEMP])
+            a += "_" + str(inx)
+            loc_addr.add(a)
```

Comment:
```suggestion
            loc_addr.add(f"{entity[CONF_TARGET_TEMP]}_{inx}")
```
Maybe this would look better, as the temporary variable `a` also doesn't add much value to what its doing

Reply:
I did that to have the following lines identical…doing as you suggest causes linebreaks in the next lines, and ugly parentheses).

Reply:
but it does avoid the extra `str()`

Reply:
Changed (including in the 2 following).